lib/event_manager/attendances/attendance.ex

```diff
@@ -0,0 +1,26 @@
+defmodule EventManager.Attendances.Attendance do
+  @moduledoc """
+  Represents an event attendance by a user account (attendee)
+  or simply by an email address.
+  """
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  @primary_key {:id, :binary_id, autogenerate: true}
+  @foreign_key_type :binary_id
+
+  schema "attendances" do
+    field :email, :string
+    belongs_to :attendee, EventManager.Users.User, foreign_key: :attendee_id
+    belongs_to :event, EventManager.Events.Event
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(attendance, attrs) do
+    attendance
+    |> cast(attrs, [:email, :event_id, :attendee_id])
+    |> validate_required([:event_id])
+  end
```

Comment:
We could actually have two different changesets, one for email-only attendances and another one for user attendance.
I was working on a prototype last week, was something along those lines:

```elixir
 def email_changeset(attendance, attrs) do
    attendance
    |> cast(attrs, [:email, :event_id, :attendee_id])
    |> validate_required([:event_id, :email])
  end

 def user_changeset(attendance, attrs) do
    attendance
    |> cast(attrs, [:email, :event_id, :attendee_id])
    |> validate_required([:event_id, :attendee_id])
  end
```

Bonus points if we manage to make `:email` and `:attendee` mutually exclusive through validations.

Reply:
@bernardini687 @apellizzn @oeN thoughts on this? Should we invest some time in properly validating this or is it better to delay this and rely on GraphQL to validate that either the email or the current user are provided?

Reply:
> We could actually have two different changesets, one for email-only attendances and another one for user attendance.
> I was working on a prototype last week, was something along those lines:
> 
> ```elixir
>  def email_changeset(attendance, attrs) do
>     attendance
>     |> cast(attrs, [:email, :event_id, :attendee_id])
>     |> validate_required([:event_id, :email])
>   end
> 
>  def user_changeset(attendance, attrs) do
>     attendance
>     |> cast(attrs, [:email, :event_id, :attendee_id])
>     |> validate_required([:event_id, :attendee_id])
>   end
> ```
> 
> Bonus points if we manage to make `:email` and `:attendee` mutually exclusive through validations.

Then,

Reply:
Then, we could pattern match to use the correct changeset, maybe something like this?
```elixir
  def create_attendance(%{email: _email} = attrs) do
    %Attendance{}
    |> Attendance.email_changeset(attrs)
    |> Repo.insert()
  end
```
And the same for `user_changeset`?

Reply:
> delay this and rely on GraphQL to validate that either the email or the current user are provided?
I dont think this would be feasible since GrapqlQL won't allow you to perform an either validation

Reply:
> Then, we could pattern match to use the correct changeset, maybe something like this?
> 
> ```elixir
>   def create_attendance(%{email: _email} = attrs) do
>     %Attendance{}
>     |> Attendance.email_changeset(attrs)
>     |> Repo.insert()
>   end
> ```
> 
> And the corresponding for `user_changeset`?

I really like this approach I think we should go this way 👍

Reply:
Plus, if the main concern is merely to avoid saving in the DB an attendance with both `email` and `attendee_id`, would it be that bad to just pop the key we don't need before insertion?

Reply:
Even something like this @bernardini687 
```
def create_attendance(attrs) do
    %Attendance{}
    |> Attendance.changeset(attrs)
    |> Repo.insert()
  end
```

And then in `Attendance.ex`
```
def changeset(attendance, %{email: email} = attrs), do: email_changeset(attendance, attrs)
def changeset(attendance, %{user: user} = attrs), do: user_changeset(attendance, attrs)
defp email_changeset(attendance, attrs) do
    attendance
    |> cast(attrs, [:email, :event_id, :attendee_id])
    |> validate_required([:event_id, :email])
  end

 defp user_changeset(attendance, attrs) do
    attendance
    |> cast(attrs, [:email, :event_id, :attendee_id])
    |> validate_required([:event_id, :attendee_id])
  end
```

Reply:
I would go for that ☝️

Reply:
I like this approac 👍 

Anyway, you can validate in GraphQL, it's a simple pattern-match in the resolver that returns an error if both email and the current user are not present. Kinda like we do for pagination params via Absinthe (if first or last are not passed, it returns an error).

But if the changeset can perform this validation and return a meaningful error to the client, then we can omit the validation at the resolver level.

Reply:
Ah ok, I was thinking you where talking about the GraphQL (language?) itself :D

Reply:
I'm going to make the discussed changes right now 👨‍🍳